blockchain/message_pool/src/msgpool.rs

```diff
@@ -815,13 +1176,22 @@ where
 
             for msg in smsgs {
                 rm(msg.from(), pending, msg.sequence(), rmsgs.borrow_mut()).await?;
+                if !repub && republished.write().await.insert(msg.cid()?) {
+                    repub = true;
+                }
             }
             for msg in msgs {
                 rm(msg.from(), pending, msg.sequence(), rmsgs.borrow_mut()).await?;
+                if !repub && republished.write().await.insert(msg.cid()?) {
+                    repub = true;
+                }
             }
         }
         *cur_tipset.write().await = Arc::new(ts);
     }
+    if repub {
+        repub_trigger.send(()).await;
```

Comment:
should this not check if the channel is full before sending? The only issue I have with this is that it's a channel of undefined length and if you send multiple republish signals, would it not do duplicate unnecessary work? Wouldn't you just want it to behave like an atomic bool and some waking mechanism, where the other thread is just notified when there was an event, perform the work, then go back to sleep?

Reply:
Only place that a repub_trigger.send() happens is in head_change, so there shouldnt be duplicate work. This is because the republish function will return early if theres nothing pending. In head_change the pending messages map will be updated to delete messages that dont need to be propitigated.